master/buildbot/db/logs.py

```diff
@@ -373,11 +373,107 @@ def thddeleteOldLogs(conn):
                 stepid_max = res_list[0]
             res.close()
 
-            # UPDATE logs SET logs.type = 'd' WHERE logs.stepid <= stepid_max AND type != 'd';
             if stepid_max:
+                if exceptions:
+                    # sqlite doesn't support multiple-table criteria in UPDATE
+                    # so we select the corresponding ids before
+
+                    # SELECT steps.id FROM steps
+                    # WHERE steps.buildid = builds.id
+                    # AND builds.builderid = builders.id
+                    # AND builders.name IN exceptions
+                    res = conn.execute(
+                        sa.select([model.steps.c.id])
+                        .where(sa.and_(model.steps.c.buildid == model.builds.c.id,
+                                       model.builds.c.builderid == model.builders.c.id,
+                                       model.builders.c.name.in_(exceptions)))
+                    )
+                    stepid_exceptions = [i[0] for i in res.fetchall()]
```

Comment:
this will store in memory all steps for all those builders from the beginning of time. This is not acceptable. I can't believe there is no way in sqlite to do this in one request. did you try subqueries?